### ISSUE TITLE
`Directolor` (web server) example not working

### ISSUE THREAD

**Issue by sfraint** [loucks1/Directolor#2]
I tried running the `Directolor` example and ran into a couple issues:

Running the code out-of-the-box - when I start the example, the radio is never able to successfully start and I get stuck in an endless loop of:
```
Attempting to start radio
Failure starting radio
```

----

I noticed that I can run the [RF24 example](https://github.com/nRF24/RF24/blob/master/examples/GettingStarted/GettingStarted.ino) without any issues, so I tried making some modifications to the `Directolor` lib and example using the working RF24 example code and passing in the working radio:
```
$ git diff
diff --git a/Directolor/Directolor.cpp b/Directolor/Directolor.cpp
index 3e19ff7..f97e96d 100644
--- a/Directolor/Directolor.cpp
+++ b/Directolor/Directolor.cpp
@@ -35,6 +35,22 @@ const uint8_t Directolor::duplicateDynamicBytesRead[DUPLICATE_CODE_UNIQUE_BYTES]
 const uint8_t Directolor::groupDynamicBytesRead[GROUP_CODE_UNIQUE_BYTES] = {3, 11, 12};
 const uint8_t Directolor::storeFavDynamicBytesRead[STORE_FAV_UNIQUE_BYTES] = {3, 10, 16, 17};
 
+Directolor::Directolor(RF24* _radio)
+{
+  radio = _radio;
+  queue = xQueueCreate(MESSAGE_QUEUE_LENGTH, sizeof(CommandItem));
+  messageIsSending = false;
+
+  xTaskCreatePinnedToCore(
+      sendCodeTask,   /* Function to implement the task */
+      "sendCodeTask", /* Name of the task */
+      10000,          /* Stack size in words */
+      NULL,           /* Task input parameter */
+      2,              /* Priority of the task */
+      NULL,           /* Task handle. */
+      0);             /* Core where the task should run */
+}
+
...
diff --git a/Directolor/Directolor.h b/Directolor/Directolor.h
index ccc6b15..d8ce0d8 100644
--- a/Directolor/Directolor.h
+++ b/Directolor/Directolor.h
@@ -66,6 +66,7 @@ class Directolor
 {
 
 public:
+    Directolor(RF24* _radio);
     Directolor(uint16_t _cepin, uint16_t _cspin, uint32_t _spi_speed = RF24_SPI_SPEED);
 
     static bool enterRemoteSearchMode();
@@ -125,7 +126,7 @@ private:
 
     static const uint8_t matchPattern[4]; // this is what we use to find out what the codes for the remote are
 
-    static RF24 radio;
+    static RF24* radio;
     static QueueHandle_t queue;
     static bool messageIsSending;
     static bool learningRemote;
@@ -596,4 +597,4 @@ private:
 
     };
 };
-#endif
\ No newline at end of file
+#endif
diff --git a/Directolor/examples/Directolor/Directolor.ino b/Directolor/examples/Directolor/Directolor.ino
index 8ad331e..52804fd 100644
--- a/Directolor/examples/Directolor/Directolor.ino
+++ b/Directolor/examples/Directolor/Directolor.ino
@@ -4,11 +4,12 @@
 #include <ESPmDNS.h>
 #include "Directolor.h"
 
-const char *ssid = "YourSSIDHere";
-const char *password = "YourPasswordHere";
+const char *ssid = "wifi_guest";
+const char *password = "password";
 
 WebServer server(80);
-Directolor directolor(22, 21);
+RF24 radio(22, 21);
+Directolor directolor(&radio);
 
 const int led = 13;
 
```
But, then when I try to issue a command through the web server, I run into `Guru Meditation Error` - `LoadStoreError`:
```
Connected to wifi_guest
IP address: 192.168.60.214
MDNS responder started
HTTP server started
Remote 1, Channel 1-Open
Open
Open
Open
Open
Guru Meditation Error: Core  1 panic'ed (LoadStoreError). Exception was unhandled.

Core  1 register dump:
PC      : 0x40147f30  PS      : 0x00060d30  A0      : 0x800d2e00  A1      : 0x3ffb1ac0  
A2      : 0x3ffc2ec4  A3      : 0x3ffb1b0f  A4      : 0x40025a46  A5      : 0x40025a77  
A6      : 0x00000055  A7      : 0x3ffb1b14  A8      : 0x00000006  A9      : 0x3ffb1b15  
A10     : 0x00000000  A11     : 0x00000013  A12     : 0x00000017  A13     : 0x0000003d  
A14     : 0x0000000c  A15     : 0x0000000f  SAR     : 0x00000008  EXCCAUSE: 0x00000003  
EXCVADDR: 0x40025a46  LBEG    : 0x40089671  LEND    : 0x40089681  LCOUNT  : 0xffffffff  


Backtrace:0x40147f2d:0x3ffb1ac00x400d2dfd:0x3ffb1af0 0x400d31a3:0x3ffb1b90 0x40147fbd:0x3ffb26d0 0x400d82cb:0x3ffb26f0 0x400d8371:0x3ffb2710 0x400d83f3:0x3ffb2750 0x400d8589:0x3ffb27b0 0x400d3ca6:0x3ffb2800 0x400dada9:0x3ffb2820 
```

----

Am I just doing something wrong? I didn't see any detailed getting-started instructions, so I just copied the 3 relevant files from this repo (Directolor `.ino`, `.cpp`, and `.h`) into a fresh Arduino IDE (1.8.19) configured for my ESP32 board.

**Comment by loucks1** [comment]
Perhaps your scenario is one that would benefit from the capacitor?  I assume you've already triple-checked your connections (I think it is reasonable because you can use the RF24 radio in the default sketch).  I'm guessing the Meditation error is coming from a dereference issue on the pointer.  I'll try your steps and see if I can make some more detailed instructions.

**Comment by sfraint** [comment]
I'm using a breakout board with an on-board 5V -> 3.3V regulator - and I believe this board has a cap on the 3.3V already. But regardless, I'm able to get the RF24 sketches working.

----

I do have an update where I was able to get reading commands and sending `c`lose commands to somewhat work. Here is what I tried and what happened:

## With `main` - not working
When I run `GetBlindCodes`, all I see is lots of repeating failure messages on the serial console:
```
Attempting to start radio
Failure starting radio
```


## With `Directolor` instance created in `setup()` - partly working

My changes (plus inserting my codes in `Directolor.h` remote 6):
```
$ git diff
diff --git a/Directolor/examples/GetBlindCodes/GetBlindCodes.ino b/Directolor/examples/GetBlindCodes/GetBlindCodes.ino
index a75e470..a3bb8a8 100644
--- a/Directolor/examples/GetBlindCodes/GetBlindCodes.ino
+++ b/Directolor/examples/GetBlindCodes/GetBlindCodes.ino
@@ -13,7 +13,7 @@
 
 #include "Directolor.h"
 
-Directolor directolor(22, 21);
+Directolor* directolor;
 
 void setup() {
 
@@ -21,10 +21,11 @@ void setup() {
   while (!Serial) {
     // some boards need to wait to ensure access to serial over USB
   }
+  directolor = new Directolor(22, 21);
 
   delay(500);
 
-  if (! directolor.enterRemoteSearchMode())
+  if (! directolor->enterRemoteSearchMode())
   {
     Serial.println("Unable to enter remote capture mode...");
   }
@@ -36,22 +37,22 @@ void loop() {
     char c = Serial.read();
     switch (c) {
       case 'd':
-        directolor.dumpCodes();
+        directolor->dumpCodes();
         break;
       case 's':
-        directolor.sendCode(6, 1, directolor_stop);
+        directolor->sendCode(6, 1, directolor_stop);
         break;
       case 'o':
-        directolor.sendCode(6, 1, directolor_open);
+        directolor->sendCode(6, 1, directolor_open);
         break;
       case 'c':
-        directolor.sendCode(6, 1, directolor_close);
+        directolor->sendCode(6, 1, directolor_close);
         break;
       case 'j':
-        directolor.sendCode(6, 1, directolor_join);
+        directolor->sendCode(6, 1, directolor_join);
         break;
       case 'r':
-        directolor.sendCode(6, 1, directolor_remove);
+        directolor->sendCode(6, 1, directolor_remove);
         break;
     }
   }
```

Now, I can read remote codes fine. However, sending codes still has some weird, but consistent issues:
1. `c`lose command makes it to the blinds, but then the serial console seems to hang for a few seconds and I see
```
E (21496) task_wdt: Task watchdog got triggered. The following tasks did not reset the watchdog in time:
E (21496) task_wdt:  - IDLE (CPU 0)
E (21496) task_wdt: Tasks currently running:
E (21496) task_wdt: CPU 0: sendCodeTask
E (21496) task_wdt: CPU 1: IDLE
E (21496) task_wdt: Aborting.

abort() was called at PC 0x400d9771 on core 0


Backtrace:0x40082e2d:0x3ffbe7bc |<-CORRUPTED
```
2. `o`pen and `s`top both immediately error out with something like this:
```
Guru Meditation Error: Core  1 panic'ed (LoadStoreError). Exception was unhandled.

Core  1 register dump:
PC      : 0x400ee0d0  PS      : 0x00060830  A0      : 0x800d1a20  A1      : 0x3ffb2730  
A2      : 0x3ffc1334  A3      : 0x3ffb277f  A4      : 0x40023eb6  A5      : 0x40023ee7  
A6      : 0x00000055  A7      : 0x3ffb2784  A8      : 0x00000006  A9      : 0x3ffb2785  
A10     : 0x00000000  A11     : 0x00000013  A12     : 0x00000017  A13     : 0x00000013  
A14     : 0x0000000c  A15     : 0x0000000f  SAR     : 0x00000008  EXCCAUSE: 0x00000003  
EXCVADDR: 0x40023eb6  LBEG    : 0x40085fe1  LEND    : 0x40085ff1  LCOUNT  : 0xffffffff  


Backtrace:0x400ee0cd:0x3ffb27300x400d1a1d:0x3ffb2760 0x400d1ca9:0x3ffb2800 0x400d3771:0x3ffb2820
```

----

At this point, I'm fairly convinced the wiring and hardware is fine, but is there any other info I can provide that will help? Or anything else I should try to troubleshoot?

**Comment by loucks1** [comment]
I'm hoping it is just a simple wiring issue.  If so, you're going too deep.  Here is the code used to initialize the radio:

`Directolor directolor(22, 21);`  (from Directolor.ino)

which calls:

`Directolor::Directolor(uint16_t _cepin, uint16_t _cspin, uint32_t _spi_speed)
{
  radio = RF24(_cepin, _cspin, _spi_speed);`

Here is the RF24 code:

`// instantiate an object for the nRF24L01 transceiver
RF24 radio(7, 8);  // using pin 7 for the CE pin, and pin 8 for the CSN pin`

You can see that by default, I'm passing 22 & 21 and RF24 uses 7 & 8

SO:  Can you confirm that you either changed Directolor.ino to match what worked in RF24 or that you correctly wired 22 & 21 to CE and CSN?

**Comment by sfraint** [comment]
Yes, I wired pins 21 and 22 to the RF24 board (and got the RF24 example working with those pins instead of the default pins).

From my follow-up comment above: with `GetBlindCodes` example I can read commands fine and I can _send a single `c`lose command_. I would assume I couldn't send or receive commands if the wiring was wrong.

----

My latest problem is that 1) close command crashes (watchdog is triggered by `sendCodeTask`) and 2) commands other than close crash with `LoadStoreError`

**Comment by loucks1** [comment]
Well, we feed the watchdog with the delay.  It is on the 2nd core.  I wonder if there is a power issue or something.  I originally started with a process loop. 

Can you try the attached?  You'll need to add a call to processLoop within your main. 

[Directolor..zip](https://github.com/loucks1/Directolor/files/9050766/Directolor.zip)

Do you have another radio or ESP32 you can try to rule out a weird hardware issue?

**Comment by alexb-dev** [comment]
An offtopic, If my esp32 has Bluetooth with BLE, do I still need a separate  NRF24L01+ ?
https://heltec.org/project/wifi-kit-32/

**Comment by sfraint** [comment]
@loucks1 I tried that zip with this main file:
```
#include "Directolor.h"

Directolor* directolor;

void setup() {

  Serial.begin(115200);
  while (!Serial) {
    // some boards need to wait to ensure access to serial over USB
  }
  directolor = new Directolor(22, 21);

  delay(500);

  if (! directolor->enterRemoteSearchMode())
  {
    Serial.println("Unable to enter remote capture mode...");
  }
}

void loop() {
  if (Serial.available())
  {
    directolor->processLoop();
    char c = Serial.read();
    switch (c) {
      case 'd':
        directolor->dumpCodes();
        break;
      case 's':
        directolor->sendCode(1, 1, directolor_stop);
        break;
      case 'o':
        directolor->sendCode(1, 1, directolor_open);
        break;
      case 'c':
        directolor->sendCode(1, 1, directolor_close);
        break;
      case 'j':
        directolor->sendCode(1, 1, directolor_join);
        break;
      case 'r':
        directolor->sendCode(1, 1, directolor_remove);
        break;
    }
  }
} // loop
```

And I don't see it reading remote codes anymore. Also, when I try to send a code, I either get `Guru Meditation Error: Core  1 panic'ed (StoreProhibited). Exception was unhandled.
` or `LoadProhibited`

**Comment by loucks1** [comment]
Can you try the new version?  Also, can you upload a large sketch to your esp32 and then reload with Directolor?  I've seen strangeness happen where loading a completely different sketch and then back to the one you want solves issues with those Errors.

**Comment by sfraint** [comment]
Sorry for the delay, I should be able to find some time this or next week to try this new version out.

Also, I might be open to swapping hardware - but at this point I'm fairly convinced the problem is a data structure/IDE/device configuration issue.

**Comment by loucks1** [comment]
Do you want me to send you a binary file that you could load manually?

**Comment by sfraint** [comment]
Sure, I can try your binary.

----

I just tried the new version compiling it myself and I still have issues. I could get it to compile and run with a few small changes (added some missing libraries, instantiating `Directolor` object in main function). However, I get watchdog errors and "CORRUPTED" backtrace when I try to `open`, `close`, or `stop`:

```
Attempting to start radioCommands available:
(search) remote - enter Remote Search Mode
(dump) codes    - dump the remote code so you can put into directolor.h (only valid if you've captured a remote code using search)
(o)pen blind    - send open code for current channel(s)
(c)lose blind   - send close code for current channel(s)
(s)top blind    - send stop code for current channel(s)
(j)oin blind    - send join code for current channel
(r)emove blind  - send remove code for current channel
(remote #)      - set remote number
(channel #)     - set channel number
(help)          - show this screen



Remote = 1, Channel = 1
Radio started
searching for remote
................................................................
Found Remote with address:  E9 93 C0
Capture Mode
bytes: 20 pipe: 1:  11 00 05 3E FF FF DF F4 86 06 67 01 00 DF F4 52 55 00 4F 58 AA Open
bytes: 20 pipe: 1:  11 00 05 0D FF FF DF F4 86 06 FF 01 00 DF F4 52 53 00 49 A9 A9 Stop
bytes: 20 pipe: 1:  11 00 05 20 FF FF DF F4 86 06 E2 01 00 DF F4 52 55 00 12 C1 BA Open
bytes: 20 pipe: 1:  11 00 05 63 FF FF DF F4 86 06 F2 01 00 DF F4 52 55 00 7F 76 AA Open
bytes: 20 pipe: 1:  11 00 05 78 FF FF DF F4 86 06 71 01 00 DF F4 52 53 00 19 CD AA Stop
bytes: 20 pipe: 1:  11 00 05 B6 FF FF DF F4 86 06 D8 01 00 DF F4 52 53 00 78 DA AA Stop
bytes: 20 pipe: 1:  11 00 05 D6 FF FF DF F4 86 06 DB 01 00 DF F4 52 44 00 24 84 AA Close
bytes: 20 pipe: 1:  11 00 05 57 FF FF DF F4 86 06 09 01 00 DF F4 52 44 00 1D B8 AA Close
bytes: 20 pipe: 1:  11 00 05 66 FF FF DF F4 86 06 27 01 00 DF F4 52 44 00 AC 5A A5 Close
Commands available:
(search) remote - enter Remote Search Mode
(dump) codes    - dump the remote code so you can put into directolor.h (only valid if you've captured a remote code using search)
(o)pen blind    - send open code for current channel(s)
(c)lose blind   - send close code for current channel(s)
(s)top blind    - send stop code for current channel(s)
(j)oin blind    - send join code for current channel
(r)emove blind  - send remove code for current channel
(remote #)      - set remote number
(channel #)     - set channel number
(help)          - show this screen


Remote = 1, Channel = 1
bytes: 12 pipe: 1:  09 10 04 4A 00 00 01 40 40 03 10 10 00 
Remote 1, Channels: 1-Open
32 55 55 55 55 55 55 55 55 55 12 F0 C0 11 00 05 A3 FF FF 78 09 86 06 91 01 00 78 09 52 55 00 75 7D
E (59749) task_wdt: Task watchdog got triggered. The following tasks did not reset the watchdog in time:
E (59749) task_wdt:  - IDLE (CPU 0)
E (59749) task_wdt: Tasks currently running:
E (59749) task_wdt: CPU 0: sendCodeTask
E (59749) task_wdt: CPU 1: loopTask
E (59749) task_wdt: Aborting.

abort() was called at PC 0x400d9d1d on core 0


Backtrace:0x40082e31:0x3ffbe7bc |<-CORRUPTED

```

**Comment by loucks1** [comment]
[GetBlindCodes.zip](https://github.com/loucks1/Directolor/files/9439975/GetBlindCodes.zip)
Attached is the non-wifi version - do you need instructions on uploading the binary file?

**Comment by sfraint** [comment]
Thanks @loucks1.

I'm not sure if I uploaded it correctly, but the binary you gave me doesn't crash! I used "verbose" mode when uploading another sketch to see the command the IDE uses to upload. Then I replaced my `Project.ino.bin` with the bin you gave me (I left the other `.bin`s alone; bootloader, partitions, boot_app0).

----

Though I'm not sure I'm using the serial console right. I tried entering `search` mode, pressing a bunch of buttons on the real remote, then `dump`ing and it didn't give me much:
```
Commands available:
(search) remote - enter Remote Search Mode
(dump) codes    - dump the remote code so you can put into directolor.h (only valid if you've captured a remote code using search)
(o)pen blind    - send open code for current channel(s)
(c)lose blind   - send close code for current channel(s)
(s)top blind    - send stop code for current channel(s)
(j)oin blind    - send join code for current channel
(r)emove blind  - send remove code for current channel
(remote #)      - set remote number
(channel #)     - set channel number
(help)          - show this screen


Remote = 1, Channel = 1
searching for remote
.............
Found Remote with address:  E9 93 C0
Capture Mode
bytes: 20 pipe: 1:  11 00 05 51 FF FF DF F4 86 06 F9 01 00 DF F4 52 55 00 7A 20 AA Open
bytes: 20 pipe: 1:  11 00 05 C8 FF FF DF F4 86 06 7A 01 00 DF F4 52 55 00 DC 11 AA Open
bytes: 20 pipe: 1:  11 00 05 2E FF FF DF F4 86 06 CF 01 00 DF F4 52 55 00 94 FB AA Open
bytes: 20 pipe: 1:  11 00 05 D0 FF FF DF F4 86 06 83 01 00 DF F4 52 44 00 F8 5E AA Close
bytes: 20 pipe: 1:  11 00 05 F8 FF FF DF F4 86 06 4C 01 00 DF F4 52 44 00 EA 58 A9 Close
bytes: 20 pipe: 1:  11 00 05 26 FF FF DF F4 86 06 7F 01 00 DF F4 52 44 00 05 C8 AA Close
bytes: 20 pipe: 1:  11 00 05 8E FF FF DF F4 86 06 AA 01 00 DF F4 52 53 00 38 21 AA Stop
bytes: 20 pipe: 1:  11 00 05 4C FF FF DF F4 86 06 EC 01 00 DF F4 52 53 00 82 AF 8E Stop
bytes: 20 pipe: 1:  11 00 0F F3 F9 43 B6 D5 AC 06 64 01 00 DF F4 57 FB 9A A1 A4 93 
bytes: 20 pipe: 1:  11 00 05 5B FF FF DF F4 86 06 5B 01 00 DF F4 52 53 00 44 8E AA Stop
/*
* Radio:  E9 93 DF F4
*/
,{ 
  {0xE9, 0x93, 0xDF, 0xF4}}

```

Then I tried sending commands, e.g. `open`, and it appeared to go into `Capture Mode`:
```
Remote 1, Channels: 1-Open
32 55 55 55 55 55 55 55 55 55 12 F0 C0 11 00 05 EB FF FF 78 09 86 06 A3 01 00 78 09 52 55 00 FD 7B
285
Capture Mode
32 55 55 55 55 55 55 55 55 55 12 F0 C0 11 00 05 28 FF FF 78 09 86 06 16 01 00 78 09 52 55 00 3F ED
285
Capture Mode
32 55 55 55 55 55 55 55 55 55 12 F0 C0 11 00 05 9C FF FF 78 09 86 06 90 01 00 78 09 52 55 00 72 07
285
Capture Mode
...
```

Am I using this binary wrong? did I upload it wrong?

**Comment by loucks1** [comment]
you have it running perfectly.  - I have the debug for the send turned on - so what starts with 32 55 55... is the code it is sending.

without compiling, you can't make it copy your remote - we'll work on compiling now that we've confirmed the hardware is working and setup properly

try this: 

Join the remote to your blind (it will overwrite your current remote)
Hold the button on the blind and press j -> your blind should jog to indicate that it is joined.
Then send open close etc to the blind and it should all work

You can duplicate the remote by putting your current remote into duplicate mode and sending a join command as well.  

Let me know if that works and we'll keep troubleshooting from there

**Comment by sfraint** [comment]
Great! I was able to join and now I can control that blind from the ESP32 using your binary.

----

I'd prefer not to modify my working remote if possible.

Any ideas how I might be able to compile this locally now?

It seems like everything points to an IDE/device configuration issue; so I'm wondering if I might have somehow gone wrong setting up my Arduino IDE? I followed the instructions from [ESP here](https://docs.espressif.com/projects/arduino-esp32/en/latest/installing.html#installing-using-arduino-ide) using the stable release link with Arduino 1.8.19. Then I tried a couple different ESP32 board types under `Tools -> Board` (including the DOIT ESP32 DEVKIT V1 and Node 32s).

Or even better, have you gotten this code to work with the native ESP-IDF?

**Comment by sfraint** [comment]
By the way, I just tried a fresh install of the Arduino IDE (1.8.19) on Ubuntu and I'm getting the same crash behavior after every command with the binary I compiled on Linux:
```
Remote = 1, Channel = 1
Radio started
Remote 1, Channels: 1-Close
32 55 55 55 55 55 55 55 55 55 12 F0 C0 11 00 05 F1 FF FF 78 09 86 06 43 01 00 78 09 52 44 00 4F 27
E (16353) task_wdt: Task watchdog got triggered. The following tasks did not reset the watchdog in time:
E (16353) task_wdt:  - IDLE (CPU 0)
E (16353) task_wdt: Tasks currently running:
E (16353) task_wdt: CPU 0: sendCodeTask
E (16353) task_wdt: CPU 1: loopTask
E (16353) task_wdt: Aborting.

abort() was called at PC 0x400dabc1 on core 0


Backtrace:0x400833d5:0x3ffbe7bc |<-CORRUPTED
```

Note: I also had the same original issue where the radio continuously failed to start until I moved the `Directolor` initialization into the `void setup()` function.

----

This is the full setup sequence I used:
* Install Arduino IDE 1.8.19
* Add ESP32 board manager prod URL
* Install `esp32` in Boards Manager
* Copy Directolor GetBlindCodes project plus Directolor files from this GitHub repo
* Install `CRC`, `RF24` and `SerialCommands` from Libraries manager
* Move directolor object creation into `setup` function

**Comment by loucks1** [comment]
I am using https://dl.espressif.com/dl/package_esp32_index.json for my board manager URL.  I'm going to bring up another computer and try yours.  I've got v1.0.6 installed in boards manager.  I'm using Node32s as the specific board type.  I have not gotten it to work via ESP-IDF.  It is going to take a couple of days on my side.

**Comment by loucks1** [comment]
I was able to recreate it with your board manager url - it seems like there is a huge difference between versions of ESP32 that we are using.  After doing some troubleshooting and trying to keep that watchdog timer from triggering, it seems that it might just be better to remove all of the ESP32 specific code (related to trying to run it on a different core) and just implement the worker portion as a simple loop callback.  The upside is that then this code should work with any board that supports RF24 and it looks like it is a good idea to stay off of core 0 anyway as that is where the WiFi and BT stuff happens.

**Comment by loucks1** [comment]
Can you try this last push?  I am using the same board as you now and resolved the issues, AFAICT.  Thanks for sticking with me through this process.

**Comment by sfraint** [comment]
That works! Thanks for fixing this!

Also, for what it's worth I tried your board manager URL and the older code worked for me. Maybe it has to do with differences in the tool dependencies in the two json files? riscv32 gcc vs xtensa gcc?

----

I do have one remaining issue and I'm happy to split it into a separate GitHub issue/close this one. It is the same thing I saw earlier: constant failure starting radio errors if I run your code as-is:
```
Attempting to start radio
Failure starting radio
```

It is easy enough to get around; if I just apply my earlier fix, your examples work for me:
```
-Directolor directolor(22, 21);
+Directolor* directolor;
```
```
 void setup() {
   Serial.begin(115200);
+  directolor = new Directolor(22, 21);
```
Then change all obj refs to pointer refs `directolor.` to `directolor->`
